# DataFrame Exercise

## By Kian Bagherlee

### Exercise 1

In [1]:
results = {}

In [2]:
import pandas as pd
import numpy as np

c:\Users\kbagh\miniconda3\Lib\site-packages\numpy\_distributor_init.py:30: UserWarning: loaded more than 1 DLL from .libs:
c:\Users\kbagh\miniconda3\Lib\site-packages\numpy\.libs\libopenblas64__v0.3.21-gcc_10_3_0.dll
c:\Users\kbagh\miniconda3\Lib\site-packages\numpy\.libs\libopenblas64__v0.3.23-246-g3d31191b-gcc_10_3_0.dll
  warnings.warn("loaded more than 1 DLL from .libs:"


In [3]:
df = pd.read_stata(
    "https://github.com/nickeubank/MIDS_Data/raw/master/US_AmericanCommunitySurvey/US_ACS_2017_10pct_sample.dta"
)
df

year  datanum   serial      cbserial          numprec subsamp  hhwt  \
0       2017        1   177686  2.017001e+12                9      64    55   
1       2017        1  1200045  2.017001e+12                6      79    25   
2       2017        1    70831  2.017000e+12  1 person record      36    57   
3       2017        1   557128  2.017001e+12                2      10    98   
4       2017        1   614890  2.017001e+12                4      96    54   
...      ...      ...      ...           ...              ...     ...   ...   
318999  2017        1   734396  2.017001e+12                4      78   100   
319000  2017        1   586263  2.017001e+12                4      57    77   
319001  2017        1   510444  2.017001e+12                2      43   152   
319002  2017        1  1220474  2.017001e+12                4      16   148   
319003  2017        1   219435  2.017000e+12                2      17    47   

                                        hhtype       cluster    adjust  ...  \
0       female householder, no husband present  2.017002e+12  1.011189  ...   
1            male householder, no wife present  2.017012e+12  1.011189  ...   
2               male householder, living alone  2.017001e+12  1.011189  ...   
3              married-couple family household  2.017006e+12  1.011189  ...   
4              married-couple family household  2.017006e+12  1.011189  ...   
...                                        ...           ...       ...  ...   
318999         married-couple family household  2.017007e+12  1.011189  ...   
319000         married-couple family household  2.017006e+12  1.011189  ...   
319001  female householder, no husband present  2.017005e+12  1.011189  ...   
319002         married-couple family household  2.017012e+12  1.011189  ...   
319003         married-couple family household  2.017002e+12  1.011189  ...   

        migcounty1                 migmet131 vetdisab  \
0                0  not in identifiable area      n/a   
1                0  not in identifiable area      n/a   
2                0  not in identifiable area      n/a   
3                0  not in identifiable area      n/a   
4                0  not in identifiable area      n/a   
...            ...                       ...      ...   
318999           0  not in identifiable area      n/a   
319000           0  not in identifiable area      n/a   
319001           0  not in identifiable area      n/a   
319002           0  not in identifiable area      n/a   
319003           0  not in identifiable area      n/a   

                         diffrem                  diffphys  \
0                            n/a                       n/a   
1        no cognitive difficulty  no ambulatory difficulty   
2       has cognitive difficulty  no ambulatory difficulty   
3        no cognitive difficulty  no ambulatory difficulty   
4                            n/a                       n/a   
...                          ...                       ...   
318999   no cognitive difficulty  no ambulatory difficulty   
319000                       n/a                       n/a   
319001  has cognitive difficulty  no ambulatory difficulty   
319002   no cognitive difficulty  no ambulatory difficulty   
319003   no cognitive difficulty  no ambulatory difficulty   

                                 diffmob diffcare  \
0                                    n/a      n/a   
1       no independent living difficulty       no   
2       no independent living difficulty       no   
3       no independent living difficulty       no   
4                                    n/a      n/a   
...                                  ...      ...   
318999  no independent living difficulty       no   
319000                               n/a      n/a   
319001  no independent living difficulty       no   
319002  no independent living difficulty       no   
319003  no independent living difficulty       no   

                               diffsens  di

### Exercise 2

In [4]:
results["ex2_num_obs"] = len(df)
len(df)

319004

### Exercise 3

In [5]:
results["ex3_num_vars"] = len(df.columns)
len(df.columns)

104

### Exercise 4

In [6]:
for c in df.columns:
    print(c)

year
datanum
serial
cbserial
numprec
subsamp
hhwt
hhtype
cluster
adjust
cpi99
region
stateicp
statefip
countyicp
countyfip
metro
city
citypop
strata
gq
farm
ownershp
ownershpd
mortgage
mortgag2
mortamt1
mortamt2
respmode
pernum
cbpernum
perwt
slwt
famunit
sex
age
marst
birthyr
race
raced
hispan
hispand
bpl
bpld
citizen
yrnatur
yrimmig
language
languaged
speakeng
hcovany
hcovpriv
hinsemp
hinspur
hinstri
hcovpub
hinscaid
hinscare
hinsva
hinsihs
school
educ
educd
gradeatt
gradeattd
schltype
degfield
degfieldd
degfield2
degfield2d
empstat
empstatd
labforce
occ
ind
classwkr
classwkrd
looking
availble
inctot
ftotinc
incwage
incbus00
incss
incwelfr
incinvst
incretir
incsupp
incother
incearn
poverty
migrate1
migrate1d
migplac1
migcounty1
migmet131
vetdisab
diffrem
diffphys
diffmob
diffcare
diffsens
diffeye
diffhear


### Exercise 5

In [7]:
df = df[["age", "inctot", "educ", "empstat"]]
df

age   inctot                       educ             empstat
0        4  9999999  nursery school to grade 4                 n/a
1       17     6000                   grade 11            employed
2       63     6150         4 years of college            employed
3       66    14000                   grade 12  not in labor force
4        1  9999999        n/a or no schooling                 n/a
...     ..      ...                        ...                 ...
318999  33    22130         4 years of college            employed
319000   4  9999999  nursery school to grade 4                 n/a
319001  20     5000                   grade 12            employed
319002  47   240000        5+ years of college            employed
319003  33    48000        5+ years of college            employed

[319004 rows x 4 columns]

### Exercise 6

In [8]:
df.sample(20)

age   inctot                 educ             empstat
622     39    75000  5+ years of college            employed
62101   11  9999999  grade 5, 6, 7, or 8                 n/a
113499  69    68000  5+ years of college          unemployed
76588   20     7000    1 year of college            employed
165740  82    31200    1 year of college  not in labor force
282786  60    32600   2 years of college            employed
77437   40    48000             grade 12            employed
318969  70    67400   4 years of college            employed
298174  57    20000   2 years of college            employed
61189   73    10000  grade 5, 6, 7, or 8  not in labor force
19602    1  9999999  n/a or no schooling                 n/a
184506  31    63000  5+ years of college            employed
302468  38        0             grade 10  not in labor force
190793  18     9600             grade 12            employed
119333  52    94000             grade 12  not in labor force
4942    49    49600             grade 12            employed
276427  58    18900    1 year of college            employed
213596  12  9999999  grade 5, 6, 7, or 8                 n/a
237130  69   126200              grade 9            employed
156817  17        0             grade 11  not in labor force

### Exercise 7

Some immediate problems I see are first, there are a lot of people with income that are the highest possible value they give, despite what the education is. Their ages tend to be either really small or really large

There are also a lot of n/a values in the Employment Status column, and are tied to the income values that are the highest possible

### Exercise 8

In [9]:
reduce_df = df[df["inctot"] != 9999999]
reduce_df

age  inctot                 educ             empstat
1       17    6000             grade 11            employed
2       63    6150   4 years of college            employed
3       66   14000             grade 12  not in labor force
5       50   50000             grade 12            employed
6       82   27100    1 year of college  not in labor force
...     ..     ...                  ...                 ...
318998  31       0  5+ years of college  not in labor force
318999  33   22130   4 years of college            employed
319001  20    5000             grade 12            employed
319002  47  240000  5+ years of college            employed
319003  33   48000  5+ years of college            employed

[265103 rows x 4 columns]

In [10]:
results["ex8_updated_num_obs"] = len(reduce_df)
len(reduce_df)

265103

### Exercise 9

In [11]:
reduce_df = reduce_df[reduce_df["empstat"] == "employed"]
reduce_df

age  inctot                 educ   empstat
1       17    6000             grade 11  employed
2       63    6150   4 years of college  employed
5       50   50000             grade 12  employed
9       17    2000             grade 12  employed
10      47   18000  n/a or no schooling  employed
...     ..     ...                  ...       ...
318995  67  125000             grade 12  employed
318999  33   22130   4 years of college  employed
319001  20    5000             grade 12  employed
319002  47  240000  5+ years of college  employed
319003  33   48000  5+ years of college  employed

[148758 rows x 4 columns]

In [12]:
results["ex9_updated_num_obs"] = len(reduce_df)
len(reduce_df)

148758

### Exercise 10

In [13]:
reduce_df["educ"].value_counts()

educ
grade 12                     47815
4 years of college           33174
1 year of college            22899
5+ years of college          20995
2 years of college           14077
grade 11                      2747
grade 5, 6, 7, or 8           2092
grade 10                      1910
n/a or no schooling           1291
grade 9                       1290
nursery school to grade 4      468
Name: count, dtype: int64

### Exercise 11

In [14]:
results["ex11_grade12_income"] = reduce_df[reduce_df["educ"] == "grade 12"][
    "inctot"
].mean()
reduce_df[reduce_df["educ"] == "grade 12"]["inctot"].mean()

38957.76068179442

### Exercise 12

In [15]:
results["ex12_college_income"] = reduce_df[reduce_df["educ"] == "4 years of college"][
    "inctot"
].mean()
reduce_df[reduce_df["educ"] == "4 years of college"]["inctot"].mean()

75485.05293301983

In [36]:
results["ex12_college_income_pct"] = (
    results["ex12_college_income"] / results["ex11_grade12_income"]
) * 100
results["ex12_college_income"] / results["ex11_grade12_income"] * 100

193.7612727527617

Our conclusion is that someone employed, whose highest education level is 4 years in college, is earning 193.76% of what someone employed, whose highest education level is grade 12, is earning.

### Exercise 13

That value suggests that you are likely to earn more money when working with a college degree compared to when working with a high school diploma.

### Exercise 14

In [28]:
reduce_df["educ"].value_counts()
high_school_dropout = [
    "grade 11",
    "grade 10",
    "grade 9",
    "grade 5, 6, 7, or 8",
    "n/a or no schooling",
    "nursery school to grade 4",
]
ex14_df = reduce_df[reduce_df["educ"].isin(high_school_dropout)]
results["ex14_high_school_dropout"] = ex14_df["inctot"].mean()
ex14_df["inctot"].mean()

26226.45692998571

### Exercise 15

In [34]:
results["ex15_grade_9"] = ex14_df[ex14_df["educ"] == "grade 9"]["inctot"].mean()
results["ex15_grade_10"] = ex14_df[ex14_df["educ"] == "grade 10"]["inctot"].mean()
results["ex15_grade_11"] = ex14_df[ex14_df["educ"] == "grade 11"]["inctot"].mean()
results["ex15_grade_12"] = reduce_df[reduce_df["educ"] == "grade 12"]["inctot"].mean()
results["ex15_4_years_of_college"] = reduce_df[
    reduce_df["educ"] == "4 years of college"
]["inctot"].mean()
results["ex15_graduate"] = reduce_df[reduce_df["educ"] == "5+ years of college"][
    "inctot"
].mean()

### Exercise 16

I believe there is a huge jump for a few reasons. One is that employers can actually see that the individual "finished" something. By seeing they fully completed high school, rather than getting close and quitting, employers feel more comfortable. Also, actually having a degree seems more appealing, as proof you did know information or took advantage.